File: Mk8.Sava.Tests/SdkCompatibility/python/compat.py
from __future__ import annotations

import base64
import os
import uuid
from datetime import datetime, timedelta, timezone
from importlib.metadata import version

from azure.storage.blob import (
    BlobBlock,
    BlobLeaseClient,
    BlobSasPermissions,
    BlobServiceClient,
    ContentSettings,
    generate_blob_sas,
)


endpoint = os.environ["MK8_SAVA_BLOB_ENDPOINT"]
account_name = os.environ["MK8_SAVA_ACCOUNT_NAME"]
account_key = os.environ["MK8_SAVA_ACCOUNT_KEY"]
service = BlobServiceClient(
    account_url=endpoint,
    credential=account_key,
    retry_total=0,
)
container_name = f"python-{uuid.uuid4().hex}"
container = service.get_container_client(container_name)
created = False

try:
    container.create_container(metadata={"runtime": "python"})
    created = True

    containers = list(
        service.list_containers(
            name_starts_with=container_name,
            include_metadata=True,
        )
    )
    assert len(containers) == 1
    assert containers[0]["name"] == container_name
    assert containers[0]["metadata"]["runtime"] == "python"

    content = bytes(
        (index * 31 + index // 263) % 251
        for index in range(256 * 1024 + 41)
    )
    block = container.get_blob_client("folder/sdk-roundtrip.bin")
    block.upload_blob(
        content,
        overwrite=True,
        content_settings=ContentSettings(
            content_type="application/x-mk8-python",
            cache_control="private,max-age=19",
        ),
        metadata={"owner": "python-sdk"},
        tags={"runtime": "python", "purpose": "compatibility"},
    )
    assert block.download_blob().readall() == content
    assert block.download_blob(offset=12_345, length=7_777).readall() == content[12_345:20_122]

    properties = block.get_blob_properties()
    assert properties.size == len(content)
    assert properties.content_settings.content_type == "application/x-mk8-python"
    assert properties.content_settings.cache_control == "private,max-age=19"
    assert properties.metadata["owner"] == "python-sdk"
    assert block.get_blob_tags()["runtime"] == "python"

    listed = list(
        container.list_blobs(
            name_starts_with="folder/",
            include=["metadata", "tags"],
        )
    )
    assert len(listed) == 1
    assert listed[0].name == block.blob_name
    assert listed[0].metadata["owner"] == "python-sdk"
    assert listed[0].tags["runtime"] == "python"

    snapshot = block.create_snapshot(metadata={"owner": "snapshot"})
    snapshot_client = container.get_blob_client(
        block.blob_name,
        snapshot=snapshot["snapshot"],
    )
    assert snapshot_client.download_blob().readall() == content
    assert snapshot_client.get_blob_properties().metadata["owner"] == "snapshot"

    staged = container.get_blob_client("staged.bin")
    block_ids = [
        base64.b64encode(value).decode("ascii")
        for value in (b"block-0001", b"block-0002")
    ]
    first = b"first-block|"
    second = b"second-block"
    staged.stage_block(block_id=block_ids[0], data=first)
    staged.stage_block(block_id=block_ids[1], data=second)
    staged.commit_block_list(
        [BlobBlock(block_id=value) for value in block_ids],
        metadata={"route": "staged"},
        tags={"runtime": "python"},
    )
    assert staged.download_blob().readall() == first + second
    committed, uncommitted = staged.get_block_list(block_list_type="committed")
    assert [item.id for item in committed] == block_ids
    assert uncommitted == []

    append = container.get_blob_client("append.log")
    append.create_append_blob()
    append.append_block(b"alpha")
    append.append_block(b"|beta")
    assert append.download_blob().readall() == b"alpha|beta"

    page = container.get_blob_client("page.bin")
    page.create_page_blob(size=1024)
    page_bytes = bytes([0x5A]) * 512
    page.upload_page(page=page_bytes, offset=0, length=len(page_bytes))
    page_ranges = list(page.list_page_ranges())
    assert len(page_ranges) == 1
    assert page_ranges[0].start == 0
    assert page_ranges[0].end == 511
    assert not page_ranges[0].cleared
    assert page.download_blob(offset=0, length=512).readall() == page_bytes

    lease = BlobLeaseClient(client=block)
    lease.acquire(lease_duration=-1)
    assert lease.id is not None
    lease.release()

    now = datetime.now(timezone.utc)
    sas = generate_blob_sas(
        account_name=account_name,
        container_name=container_name,
        blob_name=block.blob_name,
        account_key=account_key,
        permission=BlobSasPermissions(read=True),
        start=now - timedelta(minutes=1),
        expiry=now + timedelta(minutes=10),
    )
    sas_blob = block.from_blob_url(f"{block.url}?{sas}", retry_total=0)
    assert sas_blob.download_blob().readall() == content

    print(
        f"azure-storage-blob {version('azure-storage-blob')} compatibility passed "
        "for block, staged, append, page, snapshot, lease, list, tags, ranges, and SAS operations."
    )
finally:
    if created:
        try:
            container.delete_container()
        except Exception:
            pass

File: Mk8.Sava.Tests/SdkCompatibility/javascript/compat.mjs
import assert from "node:assert/strict";
import { randomUUID } from "node:crypto";
import {
  BlobClient,
  BlobSASPermissions,
  BlobServiceClient,
  StorageSharedKeyCredential,
  generateBlobSASQueryParameters,
} from "@azure/storage-blob";

const endpoint = requiredEnvironment("MK8_SAVA_BLOB_ENDPOINT");
const accountName = requiredEnvironment("MK8_SAVA_ACCOUNT_NAME");
const accountKey = requiredEnvironment("MK8_SAVA_ACCOUNT_KEY");
const sdkPackageVersion = "12.32.0";
const credential = new StorageSharedKeyCredential(accountName, accountKey);
const options = {
  retryOptions: { maxTries: 1 },
};
const service = new BlobServiceClient(endpoint, credential, options);
const containerName = `javascript-${randomUUID().replaceAll("-", "")}`;
const container = service.getContainerClient(containerName);
let created = false;

try {
  const create = await container.create({ metadata: { runtime: "node" } });
  assert.equal(create._response.status, 201);
  created = true;

  const containers = [];
  for await (const item of service.listContainers({ prefix: containerName, includeMetadata: true }))
    containers.push(item);
  assert.equal(containers.length, 1);
  assert.equal(containers[0].name, containerName);
  assert.equal(containers[0].metadata.runtime, "node");

  const content = Buffer.alloc(256 * 1024 + 37);
  for (let index = 0; index < content.length; index++)
    content[index] = (index * 29 + Math.floor(index / 257)) % 251;

  const block = container.getBlockBlobClient("folder/sdk-roundtrip.bin");
  const upload = await block.uploadData(content, {
    blobHTTPHeaders: {
      blobContentType: "application/x-mk8-javascript",
      blobCacheControl: "private,max-age=17",
    },
    metadata: { owner: "javascript-sdk" },
    tags: { runtime: "node", purpose: "compatibility" },
  });
  assert.equal(upload._response.status, 201);
  assert.deepEqual(await block.downloadToBuffer(), content);
  assert.deepEqual(await block.downloadToBuffer(12_345, 7_777), content.subarray(12_345, 20_122));

  const properties = await block.getProperties();
  assert.equal(properties.contentLength, content.length);
  assert.equal(properties.contentType, "application/x-mk8-javascript");
  assert.equal(properties.cacheControl, "private,max-age=17");
  assert.equal(properties.metadata.owner, "javascript-sdk");
  assert.equal((await block.getTags()).tags.runtime, "node");

  const listed = [];
  for await (const item of container.listBlobsFlat({
    includeMetadata: true,
    includeTags: true,
    prefix: "folder/",
  })) {
    listed.push(item);
  }
  assert.equal(listed.length, 1);
  assert.equal(listed[0].name, block.name);
  assert.equal(listed[0].metadata.owner, "javascript-sdk");
  assert.equal(listed[0].tags.runtime, "node");

  const snapshot = await block.createSnapshot({ metadata: { owner: "snapshot" } });
  const snapshotClient = block.withSnapshot(snapshot.snapshot);
  assert.deepEqual(await snapshotClient.downloadToBuffer(), content);
  assert.equal((await snapshotClient.getProperties()).metadata.owner, "snapshot");

  const staged = container.getBlockBlobClient("staged.bin");
  const blockIds = ["block-0001", "block-0002"].map(value => Buffer.from(value).toString("base64"));
  const first = Buffer.from("first-block|");
  const second = Buffer.from("second-block");
  await staged.stageBlock(blockIds[0], first, first.length);
  await staged.stageBlock(blockIds[1], second, second.length);
  await staged.commitBlockList(blockIds, {
    metadata: { route: "staged" },
    tags: { runtime: "node" },
  });
  assert.deepEqual(await staged.downloadToBuffer(), Buffer.concat([first, second]));
  assert.deepEqual(
    (await staged.getBlockList("committed")).committedBlocks.map(item => item.name),
    blockIds,
  );

  const append = container.getAppendBlobClient("append.log");
  await append.create();
  await append.appendBlock(Buffer.from("alpha"), 5);
  await append.appendBlock(Buffer.from("|beta"), 5);
  assert.equal((await append.downloadToBuffer()).toString("utf8"), "alpha|beta");

  const page = container.getPageBlobClient("page.bin");
  await page.create(1024);
  const pageBytes = Buffer.alloc(512, 0x5a);
  await page.uploadPages(pageBytes, 0, pageBytes.length);
  const pageRanges = await page.getPageRanges();
  assert.deepEqual(pageRanges.pageRange, [{ offset: 0, count: 511 }]);
  assert.deepEqual(await page.downloadToBuffer(0, 512), pageBytes);

  const lease = block.getBlobLeaseClient();
  const acquired = await lease.acquireLease(-1);
  assert.ok(acquired.leaseId);
  await lease.releaseLease();

  const sas = generateBlobSASQueryParameters(
    {
      containerName,
      blobName: block.name,
      permissions: BlobSASPermissions.parse("r"),
      startsOn: new Date(Date.now() - 60_000),
      expiresOn: new Date(Date.now() + 10 * 60_000),
    },
    credential,
  ).toString();
  const sasBlob = new BlobClient(`${block.url}?${sas}`, undefined, options);
  assert.deepEqual(await sasBlob.downloadToBuffer(), content);

  console.log(
    `@azure/storage-blob ${sdkPackageVersion} compatibility passed for block, staged, append, page, snapshot, lease, list, tags, ranges, and SAS operations.`,
  );
} finally {
  if (created)
    await container.delete().catch(() => {});
}

function requiredEnvironment(name) {
  const value = process.env[name];
  if (!value)
    throw new Error(`Missing required environment variable ${name}.`);
  return value;
}
